executor/executor.go

```diff
@@ -72,8 +72,8 @@ func (e *Executor) work() {
 type task struct {
 	id int
 	f  func() error
-	// shared are the tasks that this task is using non-exclusively.
-	shared []*task
+	// reading are the tasks that this task is using non-exclusively.
+	reading map[int]*task
```

Comment:
maybe we can have a bool flag like `accessed` to see if any keys in a task have used a key non-exclusively. this would let us keep the array approach by not adding ourselves multiple times, and is better than the map approach since we aren't doing no-ops on that action